### ISSUE TITLE
PanelStack2 'Render active panel only' not working

### ISSUE THREAD

**Issue by rygo11** [palantir/blueprint#4918]
#### Environment

- __Package version(s)__: 3.49.0 
- __Operating System__: OSX
- __Browser name and version__: Brave, Firefox, Safari

#### Code Sandbox

The issue happens on demo:
https://blueprintjs.com/docs/#core/components/panel-stack2

#### Steps to reproduce

1. Go to https://blueprintjs.com/docs/#core/components/panel-stack2
1. In demo turn off 'Render active panel only'
1. Set counter to > 0
2. open next panel
3. return to panel 1

#### Actual behavior

Counter value resets to 0 on return to panel 1

#### Expected behavior

Counter value should remain unchanged 

Note I could be wrong about how this component is supposed to work, but my understanding was that by keeping all panels mounted the state of each panel should be maintained.

**Comment by adidahiya** [comment]
At the time of writing, the core version in the docs site was 3.49.0 (note this version is important to include in your bug report, instead of simply "whatever is in the online demo", since the latter changes over time).

This seems to be a regression in PanelStack2; the counter state should be retained as you navigate through panels when `renderActivePanelOnly={false}`.

**Comment by adidahiya** [comment]
I've traced the problem to this code in PanelView2 which isolates the panel component's lifecycle from that of our PanelStack2 component:

https://github.com/palantir/blueprint/blob/beb33c2b31f139f37a325f90ae23e7847ee7522b/packages/core/src/components/panel-stack2/panelView2.tsx#L80-L88

This was added to fix bugs in controlled mode where we violated some React rules around inconsistent number of hooks called within PanelStack2. But now it seems like the memoization isn't working very well and we create a new wrapper component too often, which causes the wrapped panel to unmount and re-mount excessively. I suppose we shouldn't be using `React.useMemo` as a semantic guarantee anyway, since the [React docs](https://reactjs.org/docs/hooks-reference.html#usememo) suggest that it should only be considered a performance optimization rather than a guarantee. So we'll need to find another way to create a "PanelWrapper" which guarantees that it doesn't get re-created if a panel definition is unchanged.
